proposals/tf-operator-design-v1alpha2.md

```diff
@@ -0,0 +1,435 @@
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*
+
+- [TF-Operator Design (v1alpha2)](#tf-operator-design-v1alpha2)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [UI or API](#ui-or-api)
+  - [Design](#design)
+    - [TFController](#tfcontroller)
+    - [Distributed TensorFlow Configuration](#distributed-tensorflow-configuration)
+    - [Event-Driven](#event-driven)
+    - [Reconciler](#reconciler)
+    - [Error Handling](#error-handling)
+    - [Test](#test)
+  - [Alternatives Considered](#alternatives-considered)
+    - [Future Works](#future-works)
+    - [Related Issues](#related-issues)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+_Authors:_
+
+* @ScorpioCPH - Penghao Cen &lt;cenph@caicloud.io&gt;
+
+# TF-Operator Design (v1alpha2)
+
+## Motivation
+
+Kubeflow community currently have [tf-operator](https://github.com/kubeflow/tf-operator) (v1alpha1) for running TensorFlow jobs on Kubernetes. And we have received some refactoring [requests](https://github.com/kubeflow/tf-operator/issues?q=is%3Aissue+is%3Aopen+label%3Akind%2Fapi-change) for API changes.
+Open this file to summarize the design details and move the version of API to `v1alpha2`.
+
+## Goals
+
+- Define the structure of API `v1alpha2`.
+    + Cover most of the refactoring requests we have discussed.
+    + Simplify the API definition.
+- Define an `event-driven` mechanism for TFJob life-cycle management.
+    + And use `reconciler` mechanism as a double check.
+- Clarify the `error handing` logic.
+- Provide a `test` mechanism to verify the design and implementation.
+
+## Non-Goals
+
+- **Notes:** As we make a big change in API v1alpha2, compatibility with v1alpha1 is **NOT** be taken into consideration in this proposal.
+
+## UI or API
+
+The `TFJob` API v1alpha2 object will have the following structure:
+
+**TFJob**:
+```go
+// TFJob represents the configuration of signal TFJob
+type TFJob struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Specification of the desired behavior of the TFJob.
+    Spec TFJobSpec `json:"spec,omitempty"`
+
+    // Most recently observed status of the TFJob.
+    // This data may not be up to date.
+    // Populated by the system.
+    // Read-only.
+    Status TFJobStatus `json:"status,omitempty"`
+}
+```
+
+**TFJobSpec**:
+```go
+// TFJobSpec is a desired state description of the TFJob.
+type TFJobSpec struct {
+    // TFReplicaSpecs is map of TFReplicaType and TFReplicaSpec
+    // specifies the TF replicas to run.
+    // For example,
+    //   {
+    //     "PS": TFReplicaSpec,
+    //     "Worker": TFReplicaSpec,
+    //   }
+    TFReplicaSpecs map[TFReplicaType]*TFReplicaSpec `json:"tfReplicaSpecs"`
```

Comment:
Why not get rid of TFReplicaType and use arbitrary strings for the replica name?
The suggestion in various discussions was  to replace TFReplicaType with various properties on the Replica as opposed to inferring them based on the replica type.

As an example, restart behavior is currently inferred based on TFReplicaType. But that could instead be based on specific properties like RestartPolicy.

Reply:
Sorry, i'm not very clear about your comment: `get rid of TFReplicaType`. IMO, `TFReplicaType` is a key value for distributed TensorFlow.

Reply:
Why not just make it

```
TFReplicaSpecs map[string]*TFReplicaSpec 
```

And let the user pick what keys to use and by extension how many different replicas to have. This would allow the user to easily add a new set of replicas to do something like evaluation.

Or in the case of reinforcement learning, they might use a set of replicas to run simulations to generate training data.

Reply:
But if let user define this filed, how does tf-operator know how to hand it?

Reply:
Why does TFOperator need to do anything differently based on the name? The TFOperator should be able to handle all replicas the same way. Any differences should be controlled by properties added to the Replica.

For example, in v1alpha1 TFOperator assigns different restart policies based on TFReplicaType.

Instead of depending on the name/Type of the replica to determine the restart behavior we could add a RestartPolicy property to TFReplicaSpec.

kubeflow/tf-operator#61 is an example of an issue that would be fixed by using strings instead of TFReplicaType. If we make it a string, then a user could just call the replica "chief" or "master" based on the TF version.